snuba/query/parser/validation/functions.py

```diff
@@ -40,8 +41,38 @@ def validate(self, exp: Expression, data_source: DataSource) -> None:
 
         entity_validators: Mapping[str, FunctionCallValidator] = {}
         data_model: Optional[ColumnSet] = None
+        # If the data_source is a JoinClause it will have mutiple
+        # entities so we want to make sure we validate them all.
+        entities: List[Entity] = []
+
         if isinstance(data_source, QueryEntity):
             entity = get_entity(data_source.key)
+            entities.append(entity)
+            data_model = entity.get_data_model()
+
+        elif isinstance(data_source, JoinClause):
+            alias_map = data_source.get_alias_node_map()
+            for _, node in alias_map.items():
+                assert isinstance(node.data_source, QueryEntity)  # mypy
+                entities.append(get_entity(node.data_source.key))
+            data_model = data_source.get_columns()
+
+        elif isinstance(data_source, ProcessableQuery):
+            entity = get_entity(data_source.get_from_clause().key)
+            entities.append(entity)
+            data_model = data_source.get_columns()
+
+        else:
+            return
+
+        for entity in entities:
+            valid_funcs = entity.get_valid_function_names()
+            if exp.function_name not in valid_funcs:
+                # todo(meredith): add some metrics or logging
+                # to audit what functions are actually being used
+                # that we might have missed, later raise an err
```

Comment:
I think there is something that does not look right in this approach. Specifically in the fact that the class `FunctionCallsValidator` contain the specific logic of the individual validator (validation of the function name).
The original design of this class was that it would be executed on each and every expression in the query, it would gather whatever it needed to perform the actually validation and then delegate the specific validation logic to individual `FunctionCallValidator` classes. These ones would contain the validation logic. 
This way we can have a speartion of concern and have a single responsibility in each of these classes:
- `FunctionCallsValidator` (this class) finds the list of validation rules we need for a given expression
- Each `FunctionCallValidator` object applies one and only one validation rule.

The advantage of this is that we don't pollute the code of this class with a lot of unrelated validators and each specific function call validator rule can be unit tested in isolation thoroughly.

So in order to follow that principle we should have this class relying on a `FunctionCallValidator` (let's call it AllowedFunctionValidator) to actually do the check.

Unfortunately this design needs to change to accomplish that goal as the FunctionCallValidator only takes function parameters and not the name.

What about changing the interface of FunctionCallValidator to take the name as well together with what it needs to validate the function. (This code came before entities, and it seem we may not need to pass the data model to the validator, we could pass the entity of interest).

You may pass to the `FunctionCallValidator` these fields:
- function name
- parameters 
- entities involved.

The `AllowedFunctionValidator` would have whatever it needs to fetch the list of allowed functions.
The `SignatureValidator`, having the entity, would have what it needs to fetch the data model.

Now this class does not have to know details (like finding the data model) about how the validation is done, it jsut bothers to find the right validators to apply to a given expression.

How does it sound ?